segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookkeeperLogFactoryTests.java

```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Test
+    public void testBookkeeperClientReCreation() throws Exception {
+        BookKeeperConfig bookKeeperConfig = BookKeeperConfig.builder()
+                .with(BookKeeperConfig.BK_ACK_QUORUM_SIZE, 1)
+                .with(BookKeeperConfig.BK_LEDGER_PATH, "/ledgers")
+                .with(BookKeeperConfig.ZK_METADATA_PATH, "ledgers")
+                .with(BookKeeperConfig.BK_ENSEMBLE_SIZE, 1)
+                .with(BookKeeperConfig.BK_WRITE_QUORUM_SIZE, 1)
+                .with(BookKeeperConfig.ZK_ADDRESS, zkUtil.getZooKeeperConnectString())
+                .build();
+        RetryPolicy retryPolicy = new ExponentialBackoffRetry(1000, 3, 10000);
+        CuratorFramework zkClient = CuratorFrameworkFactory.builder()
+                .connectString(zkUtil.getZooKeeperConnectString())
+                .retryPolicy(retryPolicy)
+                .sessionTimeoutMs(30000)
+                .build();
+        zkClient.start();
+        BookKeeperLogFactory factory = new BookKeeperLogFactory(bookKeeperConfig, zkClient, this.executorService);
+        // Initialize the log, and therefore the Bookkeeper client.
+        factory.initialize();
+        // Set a timer with a longer period than the inspection period to allow client re-creation.
+        factory.getLastBookkeeperClientReset().set(new OldTimer());
+        BookKeeper oldBookkeeperClient = factory.getBookKeeperClient();
+        // Create a log the first time.
+        Assert.assertNull(factory.getLogInitializationTracker().get(0));
+        factory.createDebugLogWrapper(0);
+        // The first time we create the log the Bookkeeper client should be the same and the record for this log should
+        // be initialized.
+        Assert.assertEquals(oldBookkeeperClient, factory.getBookKeeperClient());
+        Assert.assertNotNull(factory.getLogInitializationTracker().get(0));
+        // From this point onwards, the second attempt to create the same log within the inspection period should lead
+        // to a Bookkeeper client recreation.
+        factory.createDebugLogWrapper(0);
+        Assert.assertEquals(oldBookkeeperClient, factory.getBookKeeperClient());
+        factory.createDebugLogWrapper(0);
+        Assert.assertNotEquals(oldBookkeeperClient, factory.getBookKeeperClient());
+        // Get a reference to the new Bookkeeper client.
+        oldBookkeeperClient = factory.getBookKeeperClient();
+        // The timer for this log should have been updated, so even if there are more initialization attempts, they should
+        // not lead to a new Bookkeeper client re-creation until the inspection period expires.
+        factory.createDebugLogWrapper(0);
+        Assert.assertEquals(oldBookkeeperClient, factory.getBookKeeperClient());
+        factory.createDebugLogWrapper(0);
+        Assert.assertEquals(oldBookkeeperClient, factory.getBookKeeperClient());
+    }
+
+    static class OldTimer extends Timer {
```

Comment:
Use `ManualTimer`. You can set the time as you wish.

Reply:
I tried, but this forces me to add another dependency to `segmentstore:storage:impl`, so I think that just adding this mock, tiny class has less impact to the project.